### PR TITLE
feat: Add `cum_count` expression function

### PR DESCRIPTION
**Pull request by stinodego**
Ref #13473
Ref #12420

When used without arguments, this is basically just syntactic sugar for `int_range(0, count(), 1, dtype=IndexDtype)`. 
I added this to the Rust side as we have no notion of IndexType in Python and I think it's nice to have in Rust anyway.

With arguments it's syntactic sugar for `pl.col(...).cum_count()`. It makes sense to have this along the other 'shortcut' functions we already have.

### CODE REVIEW

File: crates/polars-plan/src/dsl/mod.rs [hunk 1]
```diff
@@ -1751,6 +1751,16 @@ pub fn count() -> Expr {
     Expr::Count
 }
 
+/// Cumulative count expression.
+#[cfg(feature = "range")]
+pub fn cum_count(reverse: bool) -> Expr {
+    let mut expr = self::int_range(lit(0), count(), 1, IDX_DTYPE);
```
Comment (ritchie46) [inline crates/polars-plan/src/dsl/mod.rs:1757]: `lit(0 as IdxSize)`. Saves a cast in the optimizer.

As an aside, why `self::`? Isn't it already in scope?

**Comment by cmdlineluser** [comment]
Perhaps it could also be a good time to deprecate the `reverse=` named arg.

```python
pl.cum_count(reverse=True)
pl.cum_count().reverse()
```

(`Expr.reverse()` did not exist when `.cumcount()` was initially added - hence the need for the named arg at the time.)

**Comment by stinodego** [comment]
Totally agree. I will remove it from this PR and see if I can do the deprecation.

File: crates/polars-plan/src/dsl/mod.rs (hunk 1 above)
Comment (stinodego) [reply crates/polars-plan/src/dsl/mod.rs:1757]: The `self` was a brainfart - completely unnecessary.

**Comment by stinodego** [comment]
> Perhaps it could also be a good time to deprecate the reverse= named arg.

That's a good idea. If people want the reverse, they can do `pl.col(...).reverse().cum_count()`. I like the composability.

**Review by ritchie46 — APPROVED** [review]

**Comment by rben01** [comment]
This counts nulls as +1? Seems confusing when `df.count()` skips over nulls; I would expect the last element in the column represented by `pl.col("x").cum_count()` to be the same as `df.count().get_column("x")`. Like this:

```
┌──────┬───────────┬─────────┐
│ a    ┆ cum_count ┆ cum_len │
│ ---  ┆ ---       ┆ ---     │
│ i64  ┆ i64       ┆ i64     │
╞══════╪═══════════╪═════════╡
│ 1    ┆ 1         ┆ 1       │
│ null ┆ 1         ┆ 2       │
│ 3    ┆ 2         ┆ 3       │
└──────┴───────────┴─────────┘
```

Not sure if `cum_len` is a good name — other options include `numbered`, `row_number`, `row_index1` (or just skip it altogether and use `expr.with_row_index() + 1`) — but I would definitely consider `cum_count` misleading.

**Comment by stinodego** [comment]
I don't know what you mean. It functions exactly as you say it does - we made it match `df.count()`:

```python
import polars as pl

df = pl.DataFrame({"a": [1, 2, 3], "b": [4, None, 6]})

print(df.count())
print(df.select(pl.col("a", "b").cum_count()))
```
```
shape: (1, 2)
┌─────┬─────┐
│ a   ┆ b   │
│ --- ┆ --- │
│ u32 ┆ u32 │
╞═════╪═════╡
│ 3   ┆ 2   │
└─────┴─────┘
shape: (3, 2)
┌─────┬─────┐
│ a   ┆ b   │
│ --- ┆ --- │
│ u32 ┆ u32 │
╞═════╪═════╡
│ 1   ┆ 1   │
│ 2   ┆ 1   │
│ 3   ┆ 2   │
└─────┴─────┘
```

**Comment by rben01** [comment]
Oh, sorry never mind.. I think I got confused by the `int_range(1, count()+1, 1, dtype=IndexDtype)` part unconditionally returning row numbers. (And by the fact that elsewhere in polars, such as `group_by`, count means number of rows, not number of non-nulls.)

**Pull request merged**